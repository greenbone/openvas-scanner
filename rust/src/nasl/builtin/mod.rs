// SPDX-FileCopyrightText: 2024 Greenbone AG
//
// SPDX-License-Identifier: GPL-2.0-or-later WITH x11vnc-openssl-exception

#![doc = include_str!("README.md")]

mod array;
mod cert;
mod cryptographic;
mod description;
mod error;
mod host;
mod http;
mod isotime;
mod knowledge_base;
mod misc;
mod network;
mod preferences;
#[cfg(feature = "nasl-builtin-raw-ip")]
pub mod raw_ip;
mod regex;
mod report_functions;
mod ssh;
mod string;
mod sys;

#[cfg(test)]
mod tests;

pub use error::BuiltinError;
pub use host::HostError;
pub use knowledge_base::KBError;

use super::utils::Executor;

pub use network::socket::NaslSockets;

/// Creates a new Executor and adds all the functions to it.
pub fn nasl_std_functions() -> Executor {
    let mut executor = Executor::default();
    executor
        .add_set(array::Array)
        .add_set(report_functions::Reporting::default())
        .add_set(knowledge_base::KnowledgeBase)
        .add_set(misc::Misc)
        .add_set(string::NaslString)
        .add_set(host::Host)
        .add_set(http::NaslHttp2::default())
        .add_set(http::NaslHttp)
        .add_set(network::socket::SocketFns)
        .add_set(network::network::Network)
        .add_set(regex::RegularExpressions)
        .add_set(cryptographic::Cryptographic)
        .add_set(description::Description)
        .add_set(isotime::NaslIsotime)
        .add_set(preferences::Preferences)
        .add_set(cryptographic::rc4::CipherHandlers::default())
        .add_set(sys::Sys)
        .add_set(ssh::Ssh::default())
        .add_set(cert::NaslCerts::default());

    #[cfg(feature = "nasl-builtin-raw-ip")]
    executor.add_set(raw_ip::RawIp);
    #[cfg(feature = "nasl-builtin-raw-ip")]
    executor.add_global_vars(raw_ip::RawIp);

    executor
}
