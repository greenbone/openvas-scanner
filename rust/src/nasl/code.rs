use std::path::{Path, PathBuf};

use codespan_reporting::files::SimpleFile;

use super::{
    Loader,
    error::Level,
    syntax::{DescriptionBlock, LoadError, ParseError, Parser, Tokenizer, grammar::Ast},
};

fn parse(code: &str) -> Result<Ast, Vec<ParseError>> {
    let tokenizer = Tokenizer::tokenize(code);
    let parser = Parser::new(tokenizer);
    parser.parse_program()
}

fn parse_description_block(code: &str) -> Result<Ast, Vec<ParseError>> {
    let tokenizer = Tokenizer::tokenize(code);
    let mut parser = Parser::new(tokenizer);
    let result: Result<DescriptionBlock, ParseError> =
        parser.parse_span().map_err(|e| e.unwrap_as_spanned());
    result
        .map(|metadata| metadata.into_ast())
        .map_err(|e| vec![e])
}

pub type SourceFile = SimpleFile<String, String>;

pub struct ParseResult {
    result: Result<Ast, Vec<ParseError>>,
    file: SourceFile,
}

impl ParseResult {
    pub fn new(code: &str, path: &Path) -> Self {
        Self::new_with_parse_fn(code, path, parse)
    }

    pub fn new_with_parse_fn(
        code: &str,
        path: &Path,
        f: impl Fn(&str) -> Result<Ast, Vec<ParseError>>,
    ) -> Self {
        let file = SimpleFile::new(path.to_string_lossy().into(), code.to_owned());
        let result = f(code);
        Self { file, result }
    }

    pub fn result(self) -> Result<Ast, Vec<ParseError>> {
        self.result
    }

    pub fn num_errors(&self) -> usize {
        match &self.result {
            Ok(_) => 0,
            Err(errs) => errs.len(),
        }
    }

    pub fn emit_errors(self) -> Result<Ast, Vec<ParseError>> {
        match self.result {
            Ok(result) => Ok(result),
            Err(errors) => {
                super::error::emit_errors(&self.file, errors.iter().cloned(), Level::Error);
                Err(errors)
            }
        }
    }

    pub fn emit_errors_get_ast_and_file(self) -> Result<(Ast, SourceFile), Vec<ParseError>> {
        match self.result {
            Ok(result) => Ok((result, self.file)),
            Err(errors) => {
                super::error::emit_errors(&self.file, errors.iter().cloned(), Level::Error);
                Err(errors)
            }
        }
    }

    #[cfg(test)]
    pub fn unwrap_errors_str(self) -> String {
        let errs = self.result.unwrap_err();
        super::error::emit_errors_str(&self.file, errs.into_iter())
    }

    pub fn file(&self) -> &SourceFile {
        &self.file
    }
}

pub struct Code {
    code: String,
    path: Option<PathBuf>,
}

impl Code {
    pub fn load(loader: &dyn Loader, path: impl AsRef<Path>) -> Result<Self, LoadError> {
        Ok(Self {
            code: loader.load(&path.as_ref().to_string_lossy())?,
            path: Some(path.as_ref().to_owned()),
        })
    }

    pub fn from_string(code: &str) -> Self {
        Self {
            path: None,
            code: code.to_string(),
        }
    }

    pub fn from_string_filename(code: &str, filename: impl AsRef<Path>) -> Self {
        Self {
            code: code.to_string(),
            path: Some(filename.as_ref().to_owned()),
        }
    }

    pub fn source_file(&self) -> SourceFile {
        let path_str = self
            .path
            .as_ref()
            .map(|path| path.to_string_lossy().to_string())
            .unwrap_or("".to_string());
        SimpleFile::new(path_str, self.code().to_owned())
    }

    #[cfg(test)]
    pub fn tokenize(self) -> TokenizeResult {
        match self.path {
            Some(path) => TokenizeResult::new(&self.code, &path),
            None => unimplemented!(),
        }
    }

    pub fn parse(self) -> ParseResult {
        match self.path {
            Some(path) => ParseResult::new(&self.code, &path),
            None => ParseResult::new(&self.code, Path::new("")),
        }
    }

    pub fn parse_description_block(self) -> ParseResult {
        let path = self.path.unwrap_or(Path::new("").to_owned());
        ParseResult::new_with_parse_fn(&self.code, &path, parse_description_block)
    }

    pub fn code(&self) -> &str {
        &self.code
    }
}

#[cfg(test)]
pub use tokenize::TokenizeResult;

#[cfg(test)]
mod tokenize {
    use std::path::Path;

    use codespan_reporting::files::SimpleFile;
    use itertools::{Either, Itertools};

    use crate::nasl::{
        error::Level,
        syntax::{Token, Tokenizer, TokenizerError},
    };

    use super::{super::error, SourceFile};

    fn split_tokens_and_errors(tokenizer: Tokenizer) -> Result<Vec<Token>, Vec<TokenizerError>> {
        let (tokens, errors): (Vec<_>, Vec<_>) = tokenizer.partition_map(|a| match a {
            Ok(a) => Either::Left(a),
            Err(a) => Either::Right(a),
        });
        if !errors.is_empty() {
            Err(errors)
        } else {
            Ok(tokens)
        }
    }

    pub struct TokenizeResult {
        result: Result<Vec<Token>, Vec<TokenizerError>>,
        file: SourceFile,
    }

    impl TokenizeResult {
        pub fn new(code: &str, path: &Path) -> Self {
            let file = SimpleFile::new(path.to_string_lossy().into(), code.to_owned());
            let result = split_tokens_and_errors(Tokenizer::tokenize(code));
            Self { file, result }
        }

        pub fn result(self) -> Result<Vec<Token>, Vec<TokenizerError>> {
            self.result
        }

        pub fn emit_errors(self) -> Option<Vec<Token>> {
            match self.result {
                Ok(result) => Some(result),
                Err(errors) => {
                    error::emit_errors(&self.file, errors.into_iter(), Level::Error);
                    None
                }
            }
        }

        pub fn unwrap_errors_str(self) -> String {
            let errs = self.result.unwrap_err();
            error::emit_errors_str(&self.file, errs.into_iter())
        }
    }
}
