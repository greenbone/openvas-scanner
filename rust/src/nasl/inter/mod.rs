mod assign;
mod call;
mod declare;
mod forking_interpreter;
mod include;
mod interpreter;
mod loop_extension;
mod operator;

pub use forking_interpreter::ForkingInterpreter;
