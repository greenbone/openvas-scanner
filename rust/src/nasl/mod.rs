mod builtin;
pub mod interpreter;
pub mod syntax;
pub mod utils;

#[cfg(test)]
pub mod test_utils;

pub mod prelude {
    pub use super::builtin::ContextFactory;
    pub use super::builtin::RegisterBuilder;
    pub use super::syntax::FSPluginLoader;
    pub use super::syntax::Loader;
    pub use super::syntax::NaslValue;
    pub use super::utils::function::CheckedPositionals;
    pub use super::utils::function::FromNaslValue;
    pub use super::utils::function::Positionals;
    pub use super::utils::function::ToNaslResult;
    pub use super::utils::Context;
    pub use super::utils::ContextType;
    pub use super::utils::FunctionErrorKind;
    pub use super::utils::NaslResult;
    pub use super::utils::Register;
    pub use crate::function_set;
    pub use nasl_function_proc_macro::nasl_function;
}

pub use prelude::*;

#[cfg(test)]
pub mod test_prelude {
    pub use super::prelude::*;
    pub use super::test_utils::check_code_result;
    pub use super::test_utils::DefaultTestBuilder;
    pub use super::test_utils::TestBuilder;
    pub use crate::check_code_result_matches;
    pub use crate::check_err_matches;
}

pub use builtin::nasl_std_functions;

pub use syntax::NoOpLoader;
