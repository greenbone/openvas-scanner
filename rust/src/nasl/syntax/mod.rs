// SPDX-FileCopyrightText: 2023 Greenbone AG
//
// SPDX-License-Identifier: GPL-2.0-or-later WITH x11vnc-openssl-exception

#![doc = include_str!("README.md")]
pub mod grammar;
mod loader;
mod metadata;
mod parser;
mod token;
mod tokenizer;

pub use loader::*;
pub use metadata::DescriptionBlock;
pub use parser::{ParseError, Parser};
pub(super) use token::Ident;
pub use token::Keyword;
pub(super) use token::LiteralKind;
pub use token::Token;
pub use token::TokenKind;
pub use tokenizer::CharIndex;
pub use tokenizer::Tokenizer;
pub use tokenizer::TokenizerError;
