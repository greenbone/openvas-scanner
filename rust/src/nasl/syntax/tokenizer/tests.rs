use crate::nasl::Code;

use super::*;

pub fn tokenize_ok(file_name: &str, code: &str) -> Vec<Token> {
    let results = Code::from_string_filename(code, file_name).tokenize();
    results.emit_errors().unwrap()
}

pub fn tokenize_err(file_name: &str, code: &str) -> String {
    let results = Code::from_string_filename(code, file_name).tokenize();
    results.unwrap_errors_str()
}

macro_rules! test_ok {
    ($name: ident, $code: literal) => {
        #[test]
        fn $name() {
            insta::assert_debug_snapshot!(tokenize_ok(stringify!($name), $code));
        }
    };
}

macro_rules! test_err {
    ($name: ident, $code: literal) => {
        #[test]
        fn $name() {
            insta::assert_snapshot!(tokenize_err(stringify!($name), $code));
        }
    };
}

test_ok!(skip_whitespace, "     (       ");

test_ok!(one_char_tokens, "( ) [ ] { } . - + % ; / * : ~ & | ^");
test_ok!(
    two_char_tokens,
    "& && | || ! != !~ = == =~ > >> >= >< < << <= - -- + += ++ / /= * ** *="
);
test_ok!(three_char_tokens, ">>> >>= >!< <<=");

test_ok!(four_symbol_tokens, ">>>=");

test_ok!(string, "\"I am a closed string \"");
test_err!(unclosed_string, "\"hello I am a unclosed string\\");

test_ok!(data_string, r#"'Hello \\\'you\\\'!'"#);
test_err!(unclosed_data_string, "'Hello \\'you\\'!\\'");

test_ok!(simple_number, "1");
test_ok!(
    numbers,
    "0 0b01 1234567890 012345670 0x1234567890ABCDEF 0b02"
);
test_err!(invalid_numbers, "0x 0b 0b2");
test_err!(invalid_hex_characters, "0x123h");

test_ok!(single_line_comments, "# this is a comment\n;");

test_ok!(identifier, "help_lo _hello _h4llo 4_h4llo");

test_ok!(
    keywords,
    "for foreach if else while repeat until local_var global_var return include exit break continue"
);
test_ok!(keyword_literals, "TRUE FALSE NULL");

test_ok!(string_quoting, r"'webapps\\\\appliance\\\\'");

test_ok!(data_escape_quoting, r#"'version=\"1.0\"'"#);

test_ok!(simplified_ipv4_address, "10.187.76.12");
test_err!(wrong_ipv4_address, "10.0x 10.0.x 10.0.0.x");

test_ok!(repeat_x_times, "x() x 10;");

test_ok!(use_tokenizer, "local_var hello = 'World!';");
