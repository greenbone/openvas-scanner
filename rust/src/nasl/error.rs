use std::ops::Range;

use codespan_reporting::diagnostic::Diagnostic;
use codespan_reporting::diagnostic::Label;
use codespan_reporting::term;
use codespan_reporting::term::termcolor::Buffer;
use codespan_reporting::term::termcolor::ColorChoice;
use codespan_reporting::term::termcolor::StandardStream;

use super::code::SourceFile;
use super::syntax::CharIndex;

#[derive(Copy, Clone, Debug, PartialEq, Eq)]
pub struct Span {
    start: CharIndex,
    end: CharIndex,
}

impl Span {
    pub(crate) fn new(start: CharIndex, end: CharIndex) -> Self {
        assert!(start.0 <= end.0);
        Self { start, end }
    }

    pub(crate) fn start(&self) -> CharIndex {
        self.start
    }

    pub(crate) fn end(&self) -> CharIndex {
        self.end
    }

    pub(crate) fn join(&self, span: Span) -> Span {
        Span::new(
            CharIndex(self.start.0.min(span.start.0)),
            CharIndex(self.end.0.max(span.end.0)),
        )
    }
}

impl From<Span> for Range<usize> {
    fn from(value: Span) -> Range<usize> {
        value.start.0..value.end.0
    }
}

pub trait Spanned {
    fn span(&self) -> Span;
}

impl Spanned for Span {
    fn span(&self) -> Span {
        *self
    }
}

pub trait AsCodespanError {
    fn span(&self) -> Span;
    fn message(&self) -> String;
}

pub enum Level {
    Warn,
    Error,
}

pub fn emit_errors<T: AsCodespanError>(
    file: &SourceFile,
    errs: impl Iterator<Item = T>,
    level: Level,
) {
    let writer = StandardStream::stderr(ColorChoice::Always);
    let config = codespan_reporting::term::Config::default();
    for err in errs {
        let diagnostic = match level {
            Level::Warn => Diagnostic::warning(),
            Level::Error => Diagnostic::error(),
        };
        let diagnostic = diagnostic.with_message(err.message()).with_labels(vec![
            Label::primary((), err.span()).with_message(err.message()),
        ]);
        term::emit(&mut writer.lock(), &config, file, &diagnostic).unwrap();
    }
}

pub fn emit_errors_str<T: AsCodespanError>(
    file: &SourceFile,
    errs: impl Iterator<Item = T>,
) -> String {
    let mut writer = Buffer::no_color();
    let config = codespan_reporting::term::Config::default();
    for err in errs {
        let diagnostic = Diagnostic::error()
            .with_message(err.message())
            .with_labels(vec![
                Label::primary((), err.span()).with_message(err.message()),
            ]);
        term::emit(&mut writer, &config, file, &diagnostic).unwrap();
    }
    String::from_utf8(writer.into_inner()).unwrap()
}
