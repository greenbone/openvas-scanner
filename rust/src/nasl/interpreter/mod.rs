mod assign;
mod call;
mod declare;
mod error;
mod forking_interpreter;
mod include;
mod loop_extension;
mod nasl_value;
mod register;

#[cfg(test)]
mod tests;

use std::collections::VecDeque;

use crate::nasl::{
    Code,
    error::Span,
    syntax::{
        Ident,
        grammar::{
            Array, ArrayAccess, Atom, Binary, BinaryOperatorKind, Block, Exit, Expr, If, Include,
            Return, Statement, Unary,
        },
    },
};

use InterpreterError as Error;
use InterpreterErrorKind as ErrorKind;
use error::ExprLocation;

use super::{
    ScanCtx, ScriptCtx,
    syntax::{LiteralKind, grammar::UnaryPrefixOperatorKind},
};

pub use error::{FunctionCallError, InterpreterError, InterpreterErrorKind};
pub use forking_interpreter::ForkingInterpreter;
pub use nasl_value::NaslValue;
pub use register::Register;

pub type Result<T = NaslValue, E = Error> = std::result::Result<T, E>;

#[derive(PartialEq, Eq)]
enum InterpreterState {
    Running,
    Finished,
}

impl InterpreterState {
    fn is_finished(&self) -> bool {
        matches!(self, Self::Finished)
    }
}

/// Represents the result of a function call (`value`) along with
/// the `Span` pointing to the identifier of the function that
/// resulted in this value originally.
#[derive(Clone)]
pub struct FunctionCallData {
    value: NaslValue,
    span: Span,
}

/// This type contains the necessary data to reproduce the execution
/// flow in the case of interpreter forks.
///
/// Its two variants are
///
/// 1. `Collecting`: This variant is used by any interpreter which is
///    currently executing normally.  In this mode, the result of any
///    function call performed by the interpreter within a single
///    top-level statement will be stored along with the `Span` at
///    which the function call took place (as a safeguard). The
///    variant also stores the original `Register` and statement index
///    in order to be able to "rewind" into the exact state before the
///    statement that caused the fork.
/// 2. `Restoring`: This variant is used by all interpreters which
///    were just created due to a fork.  In this mode, the interpreter
///    will not perform normal function calls, and will instead return
///    the first value in the `data` field in place of the function
///    call. This is done until the `data` field is exhausted. At that
///    point, execution proceeds normally.
#[derive(Clone)]
pub enum ForkReentryData {
    Collecting {
        data: Vec<FunctionCallData>,
        register: Register,
        stmt_index: usize,
    },
    Restoring {
        data: VecDeque<FunctionCallData>,
    },
}

impl ForkReentryData {
    fn drain(&mut self) -> VecDeque<FunctionCallData> {
        match self {
            Self::Collecting { data, .. } => data.drain(..).collect(),
            _ => unreachable!(),
        }
    }

    fn register(&self) -> &Register {
        match self {
            Self::Collecting { register, .. } => register,
            _ => unreachable!(),
        }
    }

    fn stmt_index(&self) -> usize {
        match self {
            Self::Collecting { stmt_index, .. } => *stmt_index,
            _ => unreachable!(),
        }
    }

    fn contains_fork(&self) -> bool {
        match self {
            Self::Collecting { data, .. } => data
                .iter()
                .any(|value| matches!(value.value, NaslValue::Fork(_))),
            _ => false,
        }
    }

    /// If in `Collecting` mode, remember the given result. Otherwise
    /// do nothing.
    pub(crate) fn try_collect(&mut self, value: NaslValue, span: &Span) {
        match self {
            ForkReentryData::Collecting { data, .. } => {
                data.push(FunctionCallData { value, span: *span })
            }
            ForkReentryData::Restoring { data: _ } => {}
        }
    }

    /// If in `Restoring` mode, remove and return the first stored
    /// result from the queue. Otherwise do nothing.
    pub(crate) fn try_restore(&mut self, span: &Span) -> Result<Option<NaslValue>, Error> {
        match self {
            Self::Restoring { data } => {
                if let Some(data) = data.pop_front() {
                    if *span != data.span {
                        return Err(ErrorKind::InvalidFork.with_span(&data.span));
                    }
                    Ok(Some(data.value))
                } else {
                    Ok(None)
                }
            }
            _ => Ok(None),
        }
    }

    fn create_forks(&mut self) -> Vec<Self> {
        let mut data = vec![self.drain()];
        loop {
            let (changed, new_data) = expand_first_fork(data);
            data = new_data;
            if !changed {
                break;
            }
        }
        data.into_iter()
            .map(|data| Self::Restoring { data })
            .collect()
    }

    fn new() -> Self {
        Self::Restoring {
            data: VecDeque::new(),
        }
    }

    fn collecting(register: Register, stmt_index: usize) -> Self {
        Self::Collecting {
            data: vec![],
            register,
            stmt_index,
        }
    }

    fn is_empty_or_collecting(&self) -> bool {
        match self {
            ForkReentryData::Collecting { .. } => true,
            ForkReentryData::Restoring { data } => data.is_empty(),
        }
    }
}

/// Expand the first occurrence of `NaslValue::Fork(...)` in the list of collected function
/// calls (i.e. any called function wants to fork), by returning one list per fork value.
///
/// For example (in pseudo-code):
/// expand_first_fork([[1, Fork(2, 3)]]) = [[1, 2], [1, 3]]
/// The boolean return value is true if any expansion took place or false if
/// no expansion took place (that is, if there was no `NaslValue::Fork` in the data).
fn expand_first_fork(
    data: Vec<VecDeque<FunctionCallData>>,
) -> (bool, Vec<VecDeque<FunctionCallData>>) {
    let first_fork = data[0]
        .iter()
        .enumerate()
        .filter_map(|(index, data)| {
            let FunctionCallData { value, span } = data;
            if let NaslValue::Fork(vals) = value {
                Some((index, vals.clone(), *span))
            } else {
                None
            }
        })
        .next();
    if first_fork.is_none() {
        return (false, data);
    }
    let first_fork = first_fork.unwrap();
    let data = data
        .into_iter()
        .flat_map(|d| expand_fork_at(d, first_fork.0, first_fork.1.clone(), first_fork.2))
        .collect();
    (true, data)
}

fn expand_fork_at(
    data: VecDeque<FunctionCallData>,
    index: usize,
    vals: Vec<NaslValue>,
    span: Span,
) -> Vec<VecDeque<FunctionCallData>> {
    vals.iter()
        .map(|val| {
            let mut data = data.clone();
            data[index] = FunctionCallData {
                value: val.clone(),
                span,
            };
            data
        })
        .collect()
}

pub struct Interpreter<'ctx> {
    pub(super) register: Register,
    pub(super) scan_ctx: &'ctx ScanCtx<'ctx>,
    pub(super) script_ctx: ScriptCtx,
    pub(super) fork_reentry_data: ForkReentryData,
    stmt_index: usize,
    state: InterpreterState,
}

impl<'ctx> Interpreter<'ctx> {
    /// Creates a new Interpreter
    fn new(register: Register, scan_ctx: &'ctx ScanCtx) -> Self {
        Interpreter {
            register,
            stmt_index: 0,
            scan_ctx,
            script_ctx: ScriptCtx::default(),
            fork_reentry_data: ForkReentryData::new(),
            state: InterpreterState::Running,
        }
    }

    pub(super) async fn execute_statement(
        &mut self,
        stmt: Option<&Statement>,
    ) -> Option<Result<NaslValue, Error>> {
        self.initialize_fork_data();
        self.stmt_index += 1;
        match stmt {
            Some(stmt) => {
                let result = self.resolve(stmt).await;
                if matches!(result, Ok(NaslValue::Exit(_))) {
                    self.state = InterpreterState::Finished;
                }
                Some(result)
            }
            None => {
                self.state = InterpreterState::Finished;
                None
            }
        }
    }

    pub(super) async fn resolve(&mut self, statement: &Statement) -> Result<NaslValue, Error> {
        use Statement::*;
        match statement {
            NoOp => Ok(NaslValue::Null),
            Continue => Ok(NaslValue::Continue),
            Break => Ok(NaslValue::Break),
            ExprStmt(expr) => self.resolve_expr(expr).await,
            Block(block) => self.resolve_block(block).await,
            VarScopeDecl(var_scope_decl) => self.resolve_var_scope_decl(var_scope_decl),
            FnDecl(fn_decl) => self.resolve_fn_decl(fn_decl),
            If(if_) => self.resolve_if(if_).await,
            While(while_) => Box::pin(self.resolve_while(while_)).await,
            Repeat(repeat) => Box::pin(self.resolve_repeat(repeat)).await,
            ForEach(foreach) => Box::pin(self.resolve_foreach(foreach)).await,
            For(for_) => Box::pin(self.resolve_for(for_)).await,
            Exit(exit) => self.resolve_exit(exit).await,
            Include(include_) => self.resolve_include(include_).await,
            Return(return_) => self.resolve_return(return_).await,
        }
    }

    pub(super) async fn resolve_expr(&mut self, expr: &Expr) -> Result {
        match expr {
            Expr::Atom(atom) => Box::pin(self.resolve_atom(atom)).await,
            Expr::Binary(binary) => Box::pin(self.resolve_binary(binary)).await,
            Expr::Unary(unary) => Box::pin(self.resolve_unary(unary)).await,
            Expr::Assignment(assignment) => Box::pin(self.resolve_assignment(assignment)).await,
        }
    }

    pub async fn collect_exprs(
        &mut self,
        exprs: impl Iterator<Item = &Expr>,
    ) -> Result<Vec<NaslValue>> {
        let mut vals = vec![];
        for array_access in exprs {
            vals.push(self.resolve_expr(array_access).await?);
        }
        Ok(vals)
    }

    async fn resolve_atom(&mut self, atom: &Atom) -> Result {
        match atom {
            Atom::Literal(literal) => Ok(match &literal.kind {
                LiteralKind::String(s) => NaslValue::String(s.clone()),
                LiteralKind::Data(data) => NaslValue::Data(data.clone()),
                LiteralKind::Number(num) => NaslValue::Number(*num),
                LiteralKind::IPv4Address(ipv4_addr) => NaslValue::String(ipv4_addr.to_string()),
                LiteralKind::Null => NaslValue::Null,
                LiteralKind::Boolean(b) => NaslValue::Boolean(*b),
                LiteralKind::AttackCategory(a) => NaslValue::AttackCategory(*a),
                LiteralKind::FCTAnonArgs => NaslValue::Array(self.register.positional().to_vec()),
            }),
            Atom::FnCall(call) => self.resolve_fn_call(call).await,
            Atom::Ident(ident) => Ok(self.resolve_var(ident)?.clone()),
            Atom::Array(array) => self.resolve_array(array).await,
            Atom::ArrayAccess(array_access) => self.resolve_array_access(array_access).await,
            Atom::Increment(increment) => self.resolve_increment(increment).await,
        }
    }

    fn resolve_var(&self, ident: &Ident) -> Result {
        let var = self.register.get(ident.to_str());
        if let Some(var) = var {
            Ok(self
                .register
                .get_val(var)
                .as_value()
                .map_err(|e| e.with_span(&ident))?
                .clone())
        } else {
            Ok(NaslValue::Null)
        }
    }

    async fn resolve_array_access(&mut self, array_access: &ArrayAccess) -> Result {
        let lhs = self.resolve_expr(&array_access.lhs_expr).await?;
        let index = self.resolve_expr(&array_access.index_expr).await?;
        lhs.index(index)
            .map_err(|e| e.into_error(&array_access.lhs_expr, &array_access.index_expr))
    }

    async fn resolve_unary(&mut self, unary: &Unary) -> Result {
        let rhs = self.resolve_expr(&unary.rhs).await?;
        match unary.op.kind {
            UnaryPrefixOperatorKind::Plus => Ok(rhs),
            UnaryPrefixOperatorKind::Minus => rhs.neg(),
            UnaryPrefixOperatorKind::Bang => rhs.not(),
            UnaryPrefixOperatorKind::Tilde => rhs.bitwise_not(),
        }
        .map_err(|e| e.with_span(&*unary.rhs))
    }

    async fn resolve_binary(&mut self, binary: &Binary) -> Result {
        use BinaryOperatorKind::*;
        let lhs = self.resolve_expr(&binary.lhs).await?;
        // Short circuit
        if binary.op.kind == AmpersandAmpersand && !lhs.convert_to_boolean() {
            return Ok(NaslValue::Boolean(false));
        }
        if binary.op.kind == PipePipe && lhs.convert_to_boolean() {
            return Ok(NaslValue::Boolean(true));
        }
        let rhs = self.resolve_expr(&binary.rhs).await?;
        let eval = || match binary.op.kind {
            Plus => Ok(lhs.add(rhs)),
            Minus => Ok(lhs.sub(rhs)),
            Star => lhs.mul(rhs),
            Slash => lhs.div(rhs),
            Percent => lhs.rem(rhs),
            StarStar => lhs.pow(rhs),
            LessLess => lhs.shl(rhs),
            GreaterGreater => lhs.shr(rhs),
            GreaterGreaterGreater => lhs.shr_unsigned(rhs),
            Less => lhs.lt(rhs),
            LessEqual => lhs.le(rhs),
            Greater => lhs.gt(rhs),
            GreaterEqual => lhs.ge(rhs),
            Ampersand => lhs.bitand(rhs),
            Pipe => lhs.bitor(rhs),
            Caret => lhs.bitxor(rhs),
            AmpersandAmpersand => Ok(lhs.and(rhs)),
            PipePipe => Ok(lhs.or(rhs)),
            EqualTilde => Ok(NaslValue::Boolean(lhs.match_regex(rhs)?)),
            BangTilde => Ok(NaslValue::Boolean(!lhs.match_regex(rhs)?)),
            GreaterLess => Ok(NaslValue::Boolean(lhs.match_string(rhs)?)),
            GreaterBangLess => Ok(NaslValue::Boolean(!lhs.match_string(rhs)?)),
            EqualEqual => Ok(NaslValue::Boolean(lhs == rhs)),
            BangEqual => Ok(NaslValue::Boolean(lhs != rhs)),
        };
        eval().map_err(|err| err.into_error(&binary.lhs, &binary.rhs))
    }

    async fn resolve_array(&mut self, array: &Array) -> Result {
        Ok(NaslValue::Array(
            self.collect_exprs(array.items.items.iter()).await?,
        ))
    }

    async fn resolve_exit(&mut self, exit: &Exit) -> Result {
        let rc = Box::pin(self.resolve_expr(&exit.expr)).await?;
        match rc {
            NaslValue::Number(rc) => Ok(NaslValue::Exit(rc)),
            _ => Err(ErrorKind::NonNumericExitCode(rc).with_span(&exit.expr)),
        }
    }

    async fn resolve_return(&mut self, return_: &Return) -> Result {
        let rc = if let Some(ref expr) = return_.expr {
            self.resolve_expr(expr).await?
        } else {
            NaslValue::Null
        };
        Ok(NaslValue::Return(Box::new(rc)))
    }

    async fn resolve_if(
        &mut self,
        If {
            if_branches,
            else_branch,
        }: &If,
    ) -> Result<NaslValue, Error> {
        for (condition, block) in if_branches.iter() {
            let val = self.resolve_expr(condition).await?;
            if val.convert_to_boolean() {
                return self.resolve_block(block).await;
            }
        }
        if let Some(block) = else_branch {
            return self.resolve_block(block).await;
        }
        Ok(NaslValue::Null)
    }

    pub(crate) async fn resolve_block(&mut self, block: &Block<Statement>) -> Result {
        self.register.create_child();
        for stmt in block.items.iter() {
            match Box::pin(self.resolve(stmt)).await {
                Ok(x) => {
                    if matches!(
                        x,
                        NaslValue::Exit(_)
                            | NaslValue::Return(_)
                            | NaslValue::Break
                            | NaslValue::Continue
                    ) {
                        self.register.drop_last();
                        return Ok(x);
                    }
                }
                Err(e) => return Err(e),
            }
        }
        self.register.drop_last();
        // currently blocks return null
        Ok(NaslValue::Null)
    }

    async fn resolve_include(&mut self, include: &Include) -> Result {
        let loader = self.scan_ctx.loader();
        let code = Code::load(loader, &include.path)
            .map_err(|e| ErrorKind::LoadError(e).with_span(&include.span))?
            .parse();
        let file = code.file().clone();
        let ast = code
            .result()
            .map_err(|errs| Error::include_syntax_error(errs, file))?;
        let mut inter = ForkingInterpreter::new(ast, self.register.clone(), self.scan_ctx);
        Box::pin(inter.execute_all()).await?;
        self.register = inter.register().clone();
        Ok(NaslValue::Null)
    }

    pub(crate) fn make_forks(mut self) -> Vec<Interpreter<'ctx>> {
        let forks = self.fork_reentry_data.create_forks();
        let register = self.fork_reentry_data.register();
        let stmt_index = self.fork_reentry_data.stmt_index();
        forks
            .into_iter()
            .map(|fork| self.make_fork(fork, register, stmt_index))
            .collect()
    }

    fn make_fork(
        &self,
        fork_reentry_data: ForkReentryData,
        register: &Register,
        stmt_index: usize,
    ) -> Interpreter<'ctx> {
        Self {
            register: register.clone(),
            stmt_index,
            scan_ctx: self.scan_ctx,
            script_ctx: ScriptCtx::default(),
            fork_reentry_data,
            state: InterpreterState::Running,
        }
    }

    pub(crate) fn wants_to_fork(&self) -> bool {
        self.fork_reentry_data.contains_fork()
    }

    pub(crate) fn initialize_fork_data(&mut self) {
        if self.fork_reentry_data.is_empty_or_collecting() {
            self.fork_reentry_data =
                ForkReentryData::collecting(self.register.clone(), self.stmt_index);
        }
    }

    pub(crate) fn is_finished(&self) -> bool {
        self.state.is_finished()
    }

    pub fn register(&self) -> &Register {
        &self.register
    }
}
