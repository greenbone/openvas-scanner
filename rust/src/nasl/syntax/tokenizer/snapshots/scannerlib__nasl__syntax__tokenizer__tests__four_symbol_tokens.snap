---
source: src/nasl/syntax/tokenizer/tests.rs
expression: "tokenize_ok(stringify! (four_symbol_tokens), \">>>=\")"
---
[
    Token {
        kind: GreaterGreaterGreaterEqual,
        span: Span {
            start: CharIndex(
                0,
            ),
            end: CharIndex(
                4,
            ),
        },
    },
]
